laura/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,47 +14,69 @@
+    ArrayList<TimeRange> answer = new ArrayList<>();
 
     int lastOkStartTime = -1;
-    int intvervalLength = 0;
+    int maxOptionalAttendees = -1;
 
     for (int hour = 0; hour <= 23; hour ++) {
         for (int minute = 0; minute <= 59; minute ++) {
             int start = hour * 60 + minute;
             TimeRange potential = TimeRange.fromStartDuration(start, (int)request.getDuration());
+
+            if (!TimeRange.WHOLE_DAY.contains(potential)) 
+                break;
+
             boolean isOk = true;
+            for (String req_attendee : request.getAttendees())
+                isOk &= canAttend(events, potential, req_attendee);
 
-            if (!TimeRange.WHOLE_DAY.contains(potential)) break;
+            int optionalsCanAttend = 0;
+            for (String opt_attendee : request.getOptionalAttendees())
+                if (canAttend(events, potential, opt_attendee))
+                    optionalsCanAttend ++;
 
-            for (Event event : events) {
-                if (event.getWhen().overlaps(potential)) {
-                    for (String req_attendee : request.getAttendees()) {
-                        if (event.getAttendees().contains(req_attendee))
-                            isOk = false;
-                    }
-                }
+            /* Max # of optionals until now, then reset answer */
+            if (isOk && optionalsCanAttend > maxOptionalAttendees) {
+                maxOptionalAttendees = optionalsCanAttend;
+                answer.clear();
+                lastOkStartTime = -1;
             }
 
+            /* if less optionals can attend than best answer, ignore this time */
+            if (isOk && maxOptionalAttendees > optionalsCanAttend)
+                isOk = false;
+
             if (isOk) {
                 if (lastOkStartTime == -1) {
                     lastOkStartTime = potential.start();
-                    intvervalLength = 0;
                 }
-                intvervalLength ++;
             } else if (lastOkStartTime != -1) {
-                answer.add(TimeRange.fromStartDuration(lastOkStartTime, intvervalLength-1 + (int)request.getDuration()));
+                answer.add(TimeRange.fromStartEnd(lastOkStartTime, start - 1 + (int)request.getDuration(), false));
                 lastOkStartTime = -1;
             }
         }
     }
 
     if(lastOkStartTime != -1)
-        answer.add(TimeRange.fromStartDuration(lastOkStartTime, intvervalLength-1 + (int)request.getDuration()));
+        answer.add(TimeRange.fromStartEnd(lastOkStartTime, TimeRange.END_OF_DAY, true));
     return answer;
   }
+
+    private boolean canAttend (Collection<Event> allEvents, TimeRange potential, String attendee) {                
+        boolean canAttend = true;
+        for (Event event : allEvents) {
+            if (event.getWhen().overlaps(potential) && event.getAttendees().contains(attendee))
+                canAttend = false;
+        }
+        return canAttend;
+    }
```

Comment:
Person can't attend meeting if there is any overlap, so the method could return on the first mismatch: 

```java
    private boolean canAttend (Collection<Event> allEvents, TimeRange potential, String attendee) {                
        for (Event event : allEvents) {
            if (event.getWhen().overlaps(potential) && event.getAttendees().contains(attendee)) {
                return false;
            }
        }
        return true;
    }
```

Or, if you prefer streams:
```java
    private boolean canAttend (Collection<Event> allEvents, TimeRange potential, String attendee) {
        allEvents.stream()
          .anyMatch(
            event -> event.getWhen().overlaps(potential) && event.getAttendees().contains(attendee));
    }
```